src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,272 @@
+using chip::Mdns::TextEntry;
+
+namespace {
+
+void FreeTextEntries(MdnsService * service)
+{
+    if (service->mTextEntryies)
+    {
+        for (size_t i = 0; i < service->mTextEntrySize; i++)
+        {
+            TextEntry & entry = service->mTextEntryies[i];
+
+            if (entry.mKey)
+            {
+                chip::Platform::MemoryFree(const_cast<char *>(entry.mKey));
+                entry.mKey = nullptr;
+            }
+            if (entry.mData)
+            {
+                chip::Platform::MemoryFree(const_cast<uint8_t *>(entry.mData));
+                entry.mData = nullptr;
+            }
+        }
+        chip::Platform::MemoryFree(service->mTextEntryies);
+        service->mTextEntryies = nullptr;
+    }
+}
+
+CHIP_ERROR CopyTextEntries(MdnsService * dest, const MdnsService & source)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    if (source.mTextEntrySize)
+    {
+        dest->mTextEntryies = static_cast<TextEntry *>(chip::Platform::MemoryCalloc(source.mTextEntrySize, sizeof(TextEntry)));
+        VerifyOrExit(dest->mTextEntryies != nullptr, error = CHIP_ERROR_NO_MEMORY);
+        dest->mTextEntrySize = source.mTextEntrySize;
+
+        for (size_t i = 0; i < source.mTextEntrySize; i++)
+        {
+            TextEntry & entry = dest->mTextEntryies[i];
+            size_t keySize    = strnlen(source.mTextEntryies[i].mKey, kMdnsTextMaxSize);
+
+            VerifyOrExit(keySize < kMdnsTextMaxSize, error = CHIP_ERROR_INVALID_ARGUMENT);
+            entry.mKey = static_cast<char *>(chip::Platform::MemoryAlloc(keySize + 1));
+            VerifyOrExit(entry.mKey != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            strncpy(const_cast<char *>(entry.mKey), source.mTextEntryies[i].mKey, keySize + 1);
+            entry.mData = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(source.mTextEntryies[i].mDataSize));
+            VerifyOrExit(entry.mData != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            entry.mDataSize = source.mTextEntryies[i].mDataSize;
+            memcpy(const_cast<uint8_t *>(entry.mData), source.mTextEntryies[i].mData, entry.mDataSize);
+        }
+    }
+
+exit:
+    if (error != CHIP_NO_ERROR)
+    {
+        FreeTextEntries(dest);
+    }
+
+    return error;
+}
+
+} // namespace
+
+namespace chip {
+namespace Mdns {
+
+ServicePool::Entry::Entry() : mHasValue(false), mHasPending(false), mNodeId(0), mFabricId(0)
+{
+    mService.mTextEntryies  = nullptr;
+    mService.mTextEntrySize = 0;
+}
+
+ServicePool::Entry::Entry(Entry && rhs)
+{
+    MoveFrom(std::move(rhs));
+}
+
+ServicePool::Entry & ServicePool::Entry::operator=(Entry && rhs)
+{
+    return MoveFrom(std::move(rhs));
+}
+
+ServicePool::Entry & ServicePool::Entry::MoveFrom(Entry && rhs)
+{
+    mService    = rhs.mService;
+    mNodeId     = rhs.mNodeId;
+    mFabricId   = rhs.mFabricId;
+    mHasValue   = rhs.mHasValue;
+    mHasPending = rhs.mHasPending;
+
+    rhs.mHasValue   = false;
+    rhs.mHasPending = false;
+
```

Comment:
Should this null out `rhs.mService.mTextEntryies`, since we are taking over ownership, I assume?  The ownership semantics here need to be _really_ clearly documented.  It would make sense to me if any time `mHasValue` is false that means the entry does not point to any heap-allocated data.  Do we have that invariant now?  (I don't think we do.)  Could we have this invariant?

Reply:
Since some `MdnsService` can contain on heap-allocated text data we'll only free the data when `mHasValue && mSerivce.mTextData != nullptr`. We do have the guarantee that if `mHasValue==false` we won't touch the memory and assumes the entry to be already freed.

Reply:
> We do have the guarantee that if mHasValue==false we won't touch the memory and assumes the entry to be already freed.

This should be part of the documentation for `mHasValue`, then.

Reply:
Nulling out `rhs.mService.mTextEntryies` makes the code clearer so added a line here.